### DISCUSSION TITLE
Detecting the test failure

### DISCUSSION THREAD

**Discussion by dgumenyuk** [skhatiri/UAV-Testing-Competition#8] [Platform] [answered]
Having some test cases reaching a timeout, I'm wondering, If the timeout is reached, can it be due to a crash of a drone? Is there a way differentiate between soft and hard crash programmatically? In other words, if the timeout was reached for my test case, should I remove it from the final test suite or there is a chance it is a valuable test case that actually caused a failure of the drone?

**Comment by skhatiri** [comment]
The recent update mentioned [here](https://github.com/skhatiri/UAV-Testing-Competition#important-updates-4122023) lets you terminate the simulation with a separate timeout, and keep the flight logs until the termination time, i.e., have access to flight trajectory, plot, and minimum distance which lets you know if the drone crashed or got in an unsafe state. 
based on this, you can decide whether to include those in your final test suite or not. 

Some people have reported issues using this new update on their machines, although we have tested it successfully on our setups. But you can assume you have full access to the flight logs for such cases in our evaluation platform. 
You can also let us know if you want us to use the previous version, which does not give you any logs for such cases, to evaluate your test generator.

**Comment by dgumenyuk** [comment]
Hi! Thank you for your reply. Could you clarify where I could  find the flight logs (do you mean .ulg files)? Or I could simply call test.get_distances(), even if the test reached the timeout?

**Reply by skhatiri** [reply, marked as answer]
You should be able to get the flight logs(.ulg), plot, and distance similar to a normal execution, in the same folder and using the same methods as you mentioned.

**Reply by dgumenyuk** [reply]
Sometimes I get the following error, when running my test generator: 
Error response from daemon: Could not find the file ry in container c45c172fcb3aa2ca4f4b4babd21d842896cd0c3dd555c77cf9a6e6d885276021
aerialist.px4.docker_agent - INFO - entry - INFO - preparing the test ...Namespace(agent='local', commands='/io/mission-commands.csv', drone='14541', func=<function run_experiment at 0x7f644a51a040>, headless=True, home=None, id=None, log=None, mission='/io/mission1.plan', n=1, obstacle=[11.0, 8.0, 22.0, -7.0, 33.0, 0.0, 4.0], obstacle2=[19.0, 13.0, 15.0, 6.999999999999993, 21.0, 0.0, 23.0], obstacle3=[12.0, 13.0, 15.0, -23.0, 31.0, 0.0, 61.0], obstacle4=[], params='/io/mission-params.csv', path=None, simulator='ros', speed=1.0, test=None)
entry - INFO - setting up the test environment...
px4.simulator - INFO - Avoidance is ready (waiting 5 seconds to wrap up)
px4.drone - INFO - Conected to the drone
px4.drone - INFO - updating drone parameters
px4.drone - INFO - uploading mission /io/mission1.plan
entry - INFO - running the test...
px4.drone - INFO - **SCHEDULER STARTED** running 1 commands 
px4.drone - INFO - starting mission
px4.drone - INFO - mission started

c45c172fcb3aa2ca4f4b4babd21d842896cd0c3dd555c77cf9a6e6d885276021
testcase - INFO - test finished...
ambiegen.executors.rrt_executor - INFO - Exception during test execution
ambiegen.executors.rrt_executor - INFO - list index out of range

I set the timeout to 500. Also, the len(test.test_results) > 0 evaluates to False in this case, so I cannot get the data about my test. Is this normal or something is wrong?

**Reply by skhatiri** [reply]
Does the simulation terminate around 500s? 
I mean, is the timeout working for the execution termination, but you are not getting the logs?  or is it completely ignored?

**Reply by dgumenyuk** [reply]
Yes, the simulation finished at 500. So here is another example. Now I run the command ```SIMULATION_TIMEOUT=400 python3 cli.py generate case_studies/mission1.yaml 100```. And sometimes I get the following output:
Error response from daemon: Could not find the file ry in container a0aecf9a9c07c10909e7d0828b90d75977c0daf4bf70f81039aa47dc9dc1f42e
aerialist.px4.docker_agent - INFO - entry - INFO - preparing the test ...Namespace(agent='local', commands='/io/mission-commands.csv', drone='14541', func=<function run_experiment at 0x7f25d98ce040>, headless=True, home=None, id=None, log=None, mission='/io/mission1.plan', n=1, obstacle=[15.0, 2.0, 22.0, -3.0, 25.0, 0.0, 15.0], obstacle2=[2.0, 15.0, 21.0, 14.0, 27.0, 0.0, 88.0], obstacle3=[], obstacle4=[], params='/io/mission-params.csv', path=None, simulator='ros', speed=1.0, test=None)
entry - INFO - setting up the test environment...
px4.simulator - INFO - Avoidance is ready (waiting 5 seconds to wrap up)
px4.drone - INFO - Conected to the drone
px4.drone - INFO - updating drone parameters
px4.drone - INFO - uploading mission /io/mission1.plan
entry - INFO - running the test...
px4.drone - INFO - **SCHEDULER STARTED** running 1 commands 
px4.drone - INFO - starting mission
px4.drone - INFO - mission started

a0aecf9a9c07c10909e7d0828b90d75977c0daf4bf70f81039aa47dc9dc1f42e
testcase - INFO - test finished...
ambiegen.executors.rrt_executor - INFO - Exception during test execution
ambiegen.executors.rrt_executor - INFO - list index out of range
ambiegen.executors.abstract_executor - INFO - Executuion time: 1003.0494432449341

**Reply by dgumenyuk** [reply]
So it seems the timeout is ignored. I am using the updated version of the repo and I am running it in a host CLI. Also, here is the output from the "debug":
aerialist.px4.docker_agent - DEBUG - docker exec -it ac10e94079c4a3a67f9431967bc800c73a01177a365431eab35f457e541b4361 timeout 1000 python3 aerialist exec --drone 14541 --params '/io/mission-params.csv' --mission '/io/mission1.plan' --simulator ros --speed 1 --headless --obstacle 13.0 5.0 22.0 16.0 29.0 0 9.0 --obstacle2 9.0 8.0 21.0 -10.999999999999996 26.0 0 73.0 --obstacle3 11.0 14.0 24.0 2.0 29.0 0 23.0 --obstacle4 6.0 2.0 21.0 -29.0 34.0 0 62.0 --commands '/io/mission-commands.csv' 
I don't know why, but the timeout of 1000 is used, while I specified the timeout of 400.

**Reply by skhatiri** [reply]
This 1000 is a different timeout (DOCKER_TIMEOUT). SIMULATION_TIMEOUT happens before this if set lower. 
I will fix these kinds of ambiguities in the next versions.

**Reply by dgumenyuk** [reply]
This makes sense! But then why my simulation does not stop after 400 s if I run the following command:
```SIMULATION_TIMEOUT=400 python3 cli.py generate case_studies/mission1.yaml 100```?

**Reply by dgumenyuk** [reply]
Also, when this happens, I cannot access the logs, the test case does not contain any information about the "distances". Currently, our test generator discards such test cases, because execution ends in an “exception” block. I think there is a possibility we are missing some useful test cases because of this.
